dbms/src/Storages/KVStore/MultiRaft/Disagg/CheckpointIngestInfo.cpp

```diff
@@ -0,0 +1,258 @@
+// Copyright 2023 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <Storages/DeltaMerge/DeltaMergeStore.h>
+#include <Storages/DeltaMerge/Segment.h>
+#include <Storages/IManageableStorage.h>
+#include <Storages/KVStore/KVStore.h>
+#include <Storages/KVStore/MultiRaft/Disagg/CheckpointIngestInfo.h>
+#include <Storages/KVStore/MultiRaft/RegionPersister.h>
+#include <Storages/KVStore/TMTContext.h>
+#include <Storages/Page/PageStorage.h>
+#include <Storages/Page/V3/Universal/UniversalPageIdFormatImpl.h>
+#include <Storages/Page/V3/Universal/UniversalPageStorage.h>
+#include <Storages/Page/V3/Universal/UniversalWriteBatchImpl.h>
+#include <Storages/StorageDeltaMerge.h>
+
+namespace DB
+{
+CheckpointIngestInfo::CheckpointIngestInfo(
+    TMTContext & tmt_,
+    const TiFlashRaftProxyHelper * proxy_helper,
+    UInt64 region_id_,
+    UInt64 peer_id_)
+    : tmt(tmt_)
+    , region_id(region_id_)
+    , peer_id(peer_id_)
+    , remote_store_id(0)
+    , in_memory(false)
+    , clean_when_destruct(false)
+    , begin_time(0)
+{
+    if (!loadFromPS(proxy_helper))
+    {
+        throw Exception(
+            ErrorCodes::LOGICAL_ERROR,
+            "Failed to restore CheckpointIngestInfo, region_id={} peer_id={} store_id={}",
+            region_id,
+            peer_id,
+            tmt.getKVStore()->getStoreID(std::memory_order_relaxed));
+    }
+}
```

Comment:
I think it is better to make this function as
```
struct CheckpointIngestInfo
{
  static std::optional<CheckpointIngestInfo> restore(...);
}

std::optional<CheckpointIngestInfo> CheckpointIngestInfo::restore(...) {
  ...
  if (...) { // any failure
    return std::nullopt;
  }
  return info;
}
```
And we can remove the `in_memory` and its checks because there will be no "invalid" instances of `CheckpointIngestInfo`